File: 71. Simplify Path.cpp
class Solution {
public:
    string simplifyPath(string path) {
        deque<string> q;
        string curr;
        for (size_t i = 1; i < path.length(); i++) {
            char ch = path[i];
            if (ch == '/') {
                if (curr.empty() || curr == ".") {
                    curr.clear();
                    continue;
                }
                if (curr == "..") {
                    if (!q.empty()) q.pop_back();
                } else {
                    q.push_back(curr);
                }
                curr.clear();
            } else {
                curr.push_back(ch);
            }
        }
        if (curr == "..") {
            if (!q.empty()) q.pop_back();
        } else if (!curr.empty() && curr != ".") {
            q.push_back(curr);
        }

        string ans = "/";
        while (!q.empty()) {
            ans += q.front();
            q.pop_front();
            if (!q.empty()) ans += "/";
        }

        return ans;
    }
};

File: 71. Simplify Path.java
class Solution {
    public String simplifyPath(String path) {
        Deque<String> q = new ArrayDeque<>();
        StringBuilder curr = new StringBuilder();
        for(int i = 1; i < path.length(); i++){
            char ch = path.charAt(i);
            if(ch == '/'){
                if(curr.length() == 0 || curr.toString().equals(".")){
                    curr = new StringBuilder();
                    continue;
                }
                if(curr.toString().equals("..")){
                    if(!q.isEmpty()) q.pollLast();
                }
                else q.offerLast(curr.toString());
                curr = new StringBuilder();
            }
            else{
                curr.append(ch);
            }
        }
        if(curr.toString().equals("..")){
            if(!q.isEmpty()) q.pollLast();
        }
        else if(curr.length() != 0 && !curr.toString().equals(".")) q.offerLast(curr.toString());
        
        StringBuilder ans = new StringBuilder();
        ans.append("/");
        while(!q.isEmpty()){
            ans.append(q.pollFirst());
            if(!q.isEmpty()) ans.append("/");
        }

        return ans.toString();
    }
}  